### PR TITLE
`vim.system` fixes and improvements

### LINKED ISSUE #25000: `vim.system` command stopped with `:kill()` becomes a zombie
### Problem

Running a command with `vim.system()` and then using the object's `:kill()` method with `SIGINT` leaves the process as a zombie, per `ps`. Calling the equivalent `vim.uv.kill()` instead cleans up the process entirely the way I'd expect.

I have a CI run on my fork showing similar repro steps as below failing on all platforms, e.g. https://github.com/nojnhuh/neovim/actions/runs/6062321135/job/16448423425#step:12:3931 

The smallest change I found I can make to fix this particular error is to omit `state.handle` here to skip closing that handle: https://github.com/neovim/neovim/blob/4ea4d72af800c40511afd006ea202d008e653c3f/runtime/lua/vim/_system.lua#L33

I imagine this and the `vim.uv.kill` workaround leak the libuv handle though which is also not ideal.

cc @lewis6991 

### Steps to reproduce

```lua
nvim -l - <<EOF
local cmd = vim.system({ 'cat', '-' }, { stdin = true }) -- run forever

cmd:kill('sigint') -- fails this test
-- vim.uv.kill(cmd.pid, 'sigint') -- using this instead works

-- wait for the process not to exist
local done, err = vim.wait(5000, function()
  return vim.uv.kill(cmd.pid, 0) ~= 0
end, 100)

print(vim.fn.system('ps p ' .. tostring(cmd.pid)))

assert(done, err == -1 and 'process still exists' or 'something else went wrong')
print('ok: process is gone')
EOF
```

```
    PID TTY      STAT   TIME COMMAND
 176064 pts/3    Z+     0:00 [cat] <defunct>
E5113: Error while calling lua chunk: [string "local cmd = vim.system({ 'cat', '-' }, { stdi..."]:13: process still exists
stack traceback:
        [C]: in function 'assert'
        [string "local cmd = vim.system({ 'cat', '-' }, { stdi..."]:13: in main chunk
```

whereas the output when `vim.uv.kill()` is used instead of `:kill()`:
```
    PID TTY      STAT   TIME COMMAND
ok: process is gone
```

### Expected behavior

The process should no longer exist within a reasonable period of time after `:kill()` is invoked, where `vim.uv.kill(pid, 0)` should not return 0, like similar checks in libuv's tests: https://github.com/libuv/libuv/blob/0d78f3c758fdc41019093353a7ff4fed83005ac9/test/test-spawn.c#L105-L111.

### Neovim version (nvim -v)

v0.10.0-dev-1018+g4ea4d72af

### Vim (not Nvim) behaves the same?

No

### Operating system/version

Arch Linux, kernel 6.4.12-arch1-1

### Terminal name/version

kitty 0.29.2 + tmux 3.3_a

### $TERM environment variable

tmux-256color

### Installation

Build from source

### PR DESCRIPTION
**Pull request by lewis6991**
- Better closing of uv handles
- Improve implementation for timeout
- Some refactoring

Fixes #25000

### CODE REVIEW

File: test/functional/lua/system_spec.lua [hunk 1]
```diff
@@ -54,4 +54,21 @@ describe('vim.system', function()
     end)
   end
 
+  it ('kill processes', function()
```
Comment (zeertzjq) [inline test/functional/lua/system_spec.lua:57]:
```suggestion
  it('kill processes', function()
```

File: runtime/lua/vim/_system.lua [hunk 2]
```diff
@@ -63,10 +64,29 @@ local function new_systemobj(state)
...
 
---- @param signal integer
+local PROCESS_KILL_WAIT = 5000
+
+--- @param signal integer|string
 function SystemObj:kill(signal)
   local state = self._state
   state.handle:kill(signal)
+
+  -- wait for the process not to exist
```
Comment (nojnhuh) [inline runtime/lua/vim/_system.lua:74]: This seems to assume that any signal will result in terminating the process which may not always be true, especially for things like SIGINFO or SIGUSR1. Some processes may even need multiple SIGINTs to shut down in some cases or ignore them entirely.

Comment (lewis6991) [reply runtime/lua/vim/_system.lua:74]: Yeah I realised this too after I posted this.

The only thing I can think of is to create a luv check object to see if the pipes need closing on each event loop iteration in the case the process has terminated.

Comment (nojnhuh) [reply runtime/lua/vim/_system.lua:74]: Actually, based on some additional print statements it looks like the existing `close_handles()` call in the `on_exit` callback is already covering this, so maybe it's safe to remove the call in `:kill()`?

These are the changes I was testing with:
```diff
diff --git a/runtime/lua/vim/_system.lua b/runtime/lua/vim/_system.lua
index 6f5e95eb2..05a27a9a2 100644
--- a/runtime/lua/vim/_system.lua
+++ b/runtime/lua/vim/_system.lua
@@ -67,7 +67,6 @@ end
 function SystemObj:kill(signal)
   local state = self._state
   state.handle:kill(signal)
-  close_handles(state)
 end
 
 local MAX_TIMEOUT = 2 ^ 31
@@ -267,7 +266,9 @@ function M.run(cmd, opts, on_exit)
     detached = opts.detach,
     hide = true,
   }, function(code, signal)
+    print('in on_exit')
     close_handles(state)
+    print('handles closed')
     if state.timer then
       state.timer:stop()
       state.timer:close()
```

Comment (lewis6991) [reply runtime/lua/vim/_system.lua]: Yep, looks like `on_exit` is called when the process eventually finishes.

File: runtime/lua/vim/_system.lua [hunk 3]
```diff
@@ -2,88 +2,106 @@ local uv = vim.uv
...
-    if not handle:is_closing() then
-      handle:close()
-    end
+--- @field stdout_data? string[]
+--- @field stderr_data? string[]
+--- @field result? vim.SystemCompleted
+
+--- @enum vim.SystemSig
+local SIG = {
+  UP = 1, -- Hangup
```
Comment (justinmk) [inline runtime/lua/vim/_system.lua:35]: HUP ?

**Comment by justinmk** [comment]
Does this address or change anything about https://github.com/neovim/neovim/issues/24567#issuecomment-1666445471 ?

**Comment by lewis6991** [comment]
> Does this address or change anything about [#24567 (comment)](https://github.com/neovim/neovim/issues/24567#issuecomment-1666445471) ?

No. This just fixes handle closing and fixes some bugs with `opts.timeout`.

**Pull request merged**